### ISSUE TITLE
Switch from OpenSSL to LibreSSL ... then switch to libtls API

### ISSUE THREAD

**Issue by poolpOrg** [OpenSMTPD/OpenSMTPD#534]
This ticket is intended to stay open for a bit, it is not an overnight switch
----------------------------------------------------------------------------------------

**First, a bit of rationale**

OpenSMTPD has relied heavily on OpenSSL from day one and it has been painful all along as the daemon was designed to run chrooted with privilege separation and the OpenSSL API deals very badly with that, expecting to be able to read a file at runtime or not providing an asynchronous way to validate a certificate during a handshake.

To cope with that, OpenSMTPD like other OpenBSD daemons, shipped with a file which reimplemented the bits we needed from OpenSSL but which were lacking. This file had functions to load a private key into a buffer and work from that buffer where OpenSSL would try to load from file for instance. It does stuff it shouldn't do, because OpenSSL doesn't do it itself.

To add to the mess, the portable version of OpenSMTPD builds on various Linux distros which run various versions of OpenSSL including old ones, and which enable or not some features we rely upon: the two that pop to mind are support for AES GCM mode or TLS SNI extension. Testing the version is so brainfucked it was actually done wrong and this resulted in issues on some Linux distros just a few months ago.

We end up having a check for the OpenSSL version in our configure layer to determine if it is completely out of touch with us, then adding ifdef's at different places in the code to test if we can support this feature or if OpenSSL has been built without it... when the feature is really a no-brainer... and eventually to errors when I try to build the portable snapshot on my Linux, see it works, then someone realizes there is a missing ifdef for his particular version of OpenSSL after I publish a snapshot.

To add to the mess, LibreSSL was released recently and it is doing things right. The privilege separation enabled functions are there, there is no conditional this or that and based on a LibreSSL version we know what features are present. To avoid conflict, the functions we had reimplemented were removed from OpenBSD and therefore they end up in the compat layer of our portable branch. We now have to detect if we're running LibreSSL and don't need that shit... or if we're running OpenSSL and have to cope with the version and features mess.

Finally, LibreSSL comes with libtls which simplifies by a GREAT BUNCH the entire TLS layer by providing a very simple function set rather than forcing us to deal with the OpenSSL API and its mysterious ways (only slightly exaggerating). It's good and all, yet we can't use it because we're currently stuck with OpenSSL.

**So, what's the plan ?**

The plan is to progressively roll out of OpenSSL.

First, we stop mentioning OpenSSL in the configure layer, we WANT LibreSSL. It doesn't have to be the very last version, it has to be a version >= X where X provides the features we rely upon.

Then, we start phasing out some ifdefs, we stop supporting Linux systems that ship with a very ancient OpenSSL or with the features we rely upon disabled. If OpenSSL already fucks up on recent releases why are we even trying to work with 0.9.x releases ?? We should be pushing for a better Internet not helping distros ship old broken libraries for convenience...

When this is done, we can remove the #ifdef's that are here and there in our code, at this point we can make LibreSSL out main dependency. People running OpenSSL will still be able to rely upon it **if it is a recent OpenSSL**.

We don't intend to break OpenSMTPD overnight for OpenSSL users, we rather want it to work with a "degraded" build mode where using OpenSSL will only work if you don't build OpenSSL with random enable/disable flags and if you stick with a post-1.0.x version.

We want to push for better, just slowly enough to not fuck every one up, but enough to actually make things better rather than stay in the OpenSSL era of broken code.

**libtls**

When we have enough feedback from the Linux community that LibreSSL has been packaged, we can start relying on libtls and cleaning up the mta and smtp code to use the simpler interface.

This will be done in parallel actually but we can only switch to it once we know that we can rely fully on LibreSSL and that OpenSSL has been buried.

**Comment by poolpOrg** [comment]
I forgot to add:

I have asked around on twitter, IRC and to some users of various systems and it seems that LibreSSL is already packaged at various places. There are few remaining hosts which do not yet support LibreSSL, out of which a few are in the process of packaging.

**Comment by dfateyev** [comment]
Fedora is more conservative: https://bugzilla.redhat.com/show_bug.cgi?id=1118995
So it could cause problems with portable.

**Comment by poolpOrg** [comment]
Well in a short future I don't think it is going to cause issues, I seem to recall that the OpenSSL version in Fedora has both AES GCM and TLS SNI extension available, so the daemon would build and work as intended.

Now, I'm not going to argue why I think Tomas Mraz is wrong, it's not my call, however there is indeed a problem  at a more distant future if he insists that LibreSSL should not be packaged because more and more Linux distros are packaging it and we really really want to use libtls rather than the SSL_* API.

**Comment by dfateyev** [comment]
Well, I'm fine as long as we can use OpenSSL available in Fedora.
As for the future, I don't know, maybe Fedora technical committee will take an official decision.

**Comment by wahjava** [comment]
In FreeBSD, there is a `OPENSSL_PORT` variable that can be defined in `make.conf(5)` to override system-wide OpenSSL port to be use, since both ports conflict with each other. As long as there is OpenSMTPD stays compatibile with OpenSSL/LibreSSL, we can delegate the decision of which SSL library to use on user.

The ports anyway depend on the OpenSSL ports.

Thanks for working on this.

HTH

**Comment by poolpOrg** [comment]
Heads up:

OpenSMTPD now explicitely wants LibreSSL.

Don't worry, this does not mean that it no longer builds with OpenSSL, it simply means that the configure script no longer mentions OpenSSL and that some checks were removed to simplify the SSL library detection.

I left enough checks to ensure that OpenSSL will still work and I don't intend to break compat before a while still, but today is the day where we start seeing:

    [...]
    checking if SSL library is LibreSSL or OpenSSL... LibreSSL :-)
    checking if programs using LibreSSL functions will link... yes
    checking if programs using LibreSSL EVP_aes_256_gcm() will link... yes
    [...]

Note that most of the checks are actually to ensure that OpenSSL still works as we know LibreSSL supports them.

I have started looking at libtls and if there was a chance I could have smtpd use the old API when linked against OpenSSL and the new one when linked against LibreSSL, it is a bit too complex and I guess we can set that milestone one year away from now.

**Comment by sysfu** [comment]
OpenSMTPD 5.4.6 does not recognize the latest LibreSSL 2.2.0 libcrypto and libssl libraries in /usr/local/lib.

Forcing the issue with CFLAGS and LDFLAGS is only half successful.

```
$ sudo CFLAGS=-I/usr/local/include LDFLAGS=-L/usr/local/lib make
$ sudo make install
$ ldd /usr/sbin/smtpd
/usr/sbin/smtpd:
        Start            End              Type Open Ref GrpRef Name
        00000df21a200000 00000df21a6a9000 exe  1    0   0      /usr/sbin/smtpd
        00000df4fa561000 00000df4fa96d000 rlib 0    1   0      /usr/lib/libevent.so.4.1
        00000df4fd17f000 00000df4fd58b000 rlib 0    1   0      /usr/lib/libutil.so.12.1
        00000df453912000 00000df453d71000 rlib 0    1   0      /usr/lib/libssl.so.32.0
        00000df4b38d6000 00000df4b3ea6000 rlib 0    1   0      /usr/local/lib/libcrypto.so.33.0
        00000df4932e6000 00000df49370e000 rlib 0    1   0      /usr/lib/libm.so.9.0
        00000df5086c3000 00000df508ad8000 rlib 0    1   0      /usr/lib/libz.so.5.0
        00000df4e52e5000 00000df4e57d1000 rlib 0    1   0      /usr/lib/libc.so.78.1
        00000df4d8700000 00000df4d8700000 rtld 0    1   0      /usr/libexec/ld.so
```

If this should be broken off into its own issue let me know.

**Comment by poolpOrg** [comment]
A year has passed, the intended strategy is not workable, we have to think of a different way to deal with this.

Fortunately the successive issues in OpenSSL has forced many distros to update and we might be able to remove some ifdefs in the meantime.

**Comment by klondi** [comment]
You may be happy to know that basically any openssl < 1.0.1 is no longer maintained by the openssl team anyways and thus security releases are no longer provided.
https://www.openssl.org/policies/releasestrat.html
